From b936f962bae1fbed345238303d7fe7490002a72f Mon Sep 17 00:00:00 2001
From: Richard Levitte <levitte@openssl.org>
Date: Tue, 3 Sep 2024 19:16:05 +0200
Subject: [PATCH 1/2] util/mkinstallvars.pl: replace List::Util::pairs with out
 own

Unfortunately, List::Util::pairs didn't appear in perl core modules
before 5.19.3, and our minimum requirement is 5.10.

Fortunately, we already have a replacement implementation, and can
re-apply it in this script.

Fixes #25366
---
 util/mkinstallvars.pl | 28 ++++++++++++++++++++++++----
 1 file changed, 24 insertions(+), 4 deletions(-)

diff --git a/util/mkinstallvars.pl b/util/mkinstallvars.pl
index 52a3d607bde8b..b67a1c477fa6e 100644
--- a/util/mkinstallvars.pl
+++ b/util/mkinstallvars.pl
@@ -10,8 +10,14 @@
 # form, or passed as variable assignments on the command line.
 # The result is a Perl module creating the package OpenSSL::safe::installdata.
 
+use 5.10.0;
+use strict;
+use warnings;
+use Carp;
+
 use File::Spec;
-use List::Util qw(pairs);
+#use List::Util qw(pairs);
+sub _pairs (@);
 
 # These are expected to be set up as absolute directories
 my @absolutes = qw(PREFIX libdir);
@@ -19,9 +25,9 @@
 # as subdirectories to PREFIX or LIBDIR.  The order of the pairs is important,
 # since the LIBDIR subdirectories depend on the calculation of LIBDIR from
 # PREFIX.
-my @subdirs = pairs (PREFIX => [ qw(BINDIR LIBDIR INCLUDEDIR APPLINKDIR) ],
-                     LIBDIR => [ qw(ENGINESDIR MODULESDIR PKGCONFIGDIR
-                                    CMAKECONFIGDIR) ]);
+my @subdirs = _pairs (PREFIX => [ qw(BINDIR LIBDIR INCLUDEDIR APPLINKDIR) ],
+                      LIBDIR => [ qw(ENGINESDIR MODULESDIR PKGCONFIGDIR
+                                     CMAKECONFIGDIR) ]);
 # For completeness, other expected variables
 my @others = qw(VERSION LDLIBS);
 
@@ -151,3 +157,17 @@ package OpenSSL::safe::installdata;
 
 1;
 _____
+
+######## Helpers
+
+sub _pairs (@) {
+    croak "Odd number of arguments" if @_ & 1;
+
+    my @pairlist = ();
+
+    while (@_) {
+        my $x = [ shift, shift ];
+        push @pairlist, $x;
+    }
+    return @pairlist;
+}

From 320dfc2b90b0fa605e0a91c7de519289b3b03014 Mon Sep 17 00:00:00 2001
From: Richard Levitte <levitte@openssl.org>
Date: Wed, 4 Sep 2024 06:56:32 +0200
Subject: [PATCH 2/2] fixup! util/mkinstallvars.pl: replace List::Util::pairs
 with out own

---
 util/mkinstallvars.pl | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/util/mkinstallvars.pl b/util/mkinstallvars.pl
index b67a1c477fa6e..950edf173382e 100644
--- a/util/mkinstallvars.pl
+++ b/util/mkinstallvars.pl
@@ -160,6 +160,15 @@ package OpenSSL::safe::installdata;
 
 ######## Helpers
 
+# _pairs LIST
+#
+# This operates on an even-sized list, and returns a list of "ARRAY"
+# references, each containing two items from the given LIST.
+#
+# It is a quick cheap reimplementation of List::Util::pairs(), a function
+# we cannot use, because it only appeared in perl v5.19.3, and we claim to
+# support perl versions all the way back to v5.10.
+
 sub _pairs (@) {
     croak "Odd number of arguments" if @_ & 1;
 
